docs/source/features/optimizations/sequence_packing.rst

```diff
@@ -1,5 +1,5 @@
-Throughput Optimizations
-========================
+Sequence Packing
```

Comment:
Technical edit of sequence_packing.rst by line numbers.

Line 3. Need to add summary text for this topic. Something like:

Sequence packing in Large Language Models (LLMs) is a technique used to improve the efficiency of training by better managing variable-length sequences and optimizing computational resources. This section describes sequence packing for Supervised Fine-Tuning (SFT) and Parameter-Efficient Fine-Tuning (PERT) methods and provides instructions for running it.

Lines 7 - 8. Delete "Overview" and ^^^^^^^^ characters. This heading is not needed.

Lines 10 - 14 fix punctuation.

When fine-tuning a LLM with either SFT or PEFT, GPU
underutilization is a common problem due to an inefficient data pipeline. This problem occurs because most fine-tuning datasets have a skewed distribution of sequence lengths, with many short sequences and a few long sequences, following Zipf’s Law.

Transformer models can only take in fixed length inputs, so the input has to be padded with many unused pad tokens. This method is inefficient in two ways:

Lines 24 - 25. fix punctuation.

While sequences for pretraining can be concatenated naively, this is not the case for SFT and instruction fine-tuning, where each input sequence should be treated individually. 

Line 30. fix spelling.

kernels in Flash Attention and Transformer Engine. With this, attention values between sequences are never calculated,

Lines 34 - 38. revise.

All things considered, the NeMo Framework implementation of sequence packing provides: 

- Up to 10X performance improvement in terms of FLOPs.
- Up to 6X performance improvement in terms of training time.
- No impact on model convergence.

Lines 42 - 43. Use imperative verb.

Run SFT/PEFT with Packed Sequence
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Line 44. Add lead-in text for this section. suggested addition.

This section provides procedures for preparing the dataset and adjusting the training configuration for SFT and PEFT methods.

Lines 57 - 59. suggested revision.

1. The online processing code in GPTSFTDataset is run. This process includes operations such as prompt template manipulation, sequence length truncation, and tokenization. The result is an array of tokenized sequences represented by indices.

2. The sequences are grouped by length and a packing algorithm is run.

Lines 62 - 65. fix punctuation and syntax. 

Currently, two variants of first fit are supported:

* ``first_fit_decreasing`` sorts the sequences in decreasing order before applying the first-fit algorithm. It generates a more optimal packing, but it tends to keep all short sequences together, which may have an impact for convergence.

* ``first_fit_shuffle`` runs first-fit in a random order. Packing is less optimal but it keeps the dataset order random. The recommendation is to run first_fit_shuffle and check the packed sequence lengths. 

If they are similar to the target length (i.e. efficient packing), then use ``first_fit_shuffle``. Otherwise try ``first_fit_decreasing``.

Lines 80 - 93. suggested revision

..note::
  *  If your model or dataset requires non-default configs for conventional SFT/PEFT training in NeMo, you will need to pass in the same configs to ``model.data.train_ds`` as you would for training with unpacked dataset.

  * ``model.data.train_ds.max_seq_length`` is the length to truncate each sequence before packing multiple sequences to the size of packed sequence (``pack_size``). ``max_seq_length`` should be set to the same value as unpacked data, and can be determined by examining the distribution of sequence lengths in the dataset.
    
  *  ``pack_sizes`` is a list of packed sequence lengths. In this example, there will be three output files, one for each pack size. The output files are named ``<output_folder>/packed_{pack_size}_seed{seed}.npy``. This argument is a list because you will likely want to experiment with a few ``pack_sizes`` to find out which length can fill the GPU memory without exceeding it. Adjusting ``pack_size`` is analogous to adjusting the micro batch size in the unpacked case.

Line 99. fix punctuation.

To train with packed sequences, you need to change four items in the SFT/PEFT config file.

Line 101. revise.

1. Enable the packed_sequence flag:

Line 107. fix punctuation.

2. Use the new dataset file instead of the original jsonl file:

Line 109. fix punctuation.

4. Adjust the batch sizes:

Line 133. revise.

Now, you are all set to fine-tune your model with a much improved throughput!


Lines 138 - 139. revise. Also, verify that the link works in the rendered HTML file.

Sequence packing in NeVA, which is a part of Multimodal Language Models (LLMs), differs from the SFT/PEFT approach. For details,
please refer to :doc:`../multimodal/mllm/sequence_packing`.